src/v/archival/archiver_manager.cc

```diff
@@ -0,0 +1,1048 @@
+            }
+        }
+    };
+
+    struct tr_active_to_stopping {
+        template<class Event>
+        void operator()(
+          const Event&,
+          state_machine_t& fsm,
+          st_active& active,
+          st_stopping_async& stopping) {
+            vlog(fsm._ctxlog.info, "stopping ntp_archiver");
+            stopping.reset_units = fsm._part->get_archiver_reset_units();
+            auto completion = [&fsm](ss::future<> fut) {
+                if (fut.failed()) {
+                    vlog(
+                      fsm._ctxlog.error,
+                      "ntp_archiver shutdown failed {}",
+                      fut.get_exception());
+                    fsm.process_event(
+                      ev_archiver_failure{.error = fut.get_exception()});
+                } else {
+                    fsm.process_event(ev_archiver_stopped{});
+                }
+            };
+            ssx::background = fsm.bg_stop_and_dispose_archiver(active.archiver)
+                                .then_wrapped(completion)
+                                .finally([arch = active.archiver] {});
+            active.clear();
+        }
+    };
+
+    struct tr_stopping_to_passive {
+        void operator()(
+          const ev_archiver_stopped&,
+          state_machine_t& fsm,
+          st_stopping_async& prev,
+          st_passive& next) {
+            vlog(fsm._ctxlog.info, "ntp_archiver stopped");
+            prev.clear();
+            if (next.trigger) {
+                next.trigger->set_value();
+            }
+        }
+        void operator()(
+          const ev_archiver_failure& f,
+          state_machine_t& fsm,
+          st_stopping_async& prev,
+          st_passive& next) {
+            vlog(
+              fsm._ctxlog.error, "ntp_archiver failed to shutdown {}", f.error);
+            prev.clear();
+            if (next.trigger) {
+                next.trigger->set_exception(f.error);
+            }
+        }
+    };
+
+    template<class Src, class Event, class Target, class Action>
+    using row = boost::msm::front::Row<Src, Event, Target, Action>;
+    using none = boost::msm::front::none;
+    using defer = boost::msm::front::Defer;
+
+    // clang-format off
+    // The state transition table is used to define a transition graph for the
+    // FSM in a declarative manner. Boost.MSM uses graph_mpl library to validate
+    // the transition graph at compile time. It checks that all states are
+    // reachable. The compile time checks are disabled by default. To enable
+    // them one has to use 'msm::back::mpl_graph_fsm_check' backend defined in
+    // 'boost/msm/back/mpl_graph_fsm_check.hpp'. It's too expensive in terms of
+    // compile time. Authors of the BoostMPL recommend to use it only when the
+    // state transition table changes.
+    using transition_table = boost::mpl::vector<
+    //  Source state      | Event                  | Dest state        | Action
```

Comment:
There are four states, not two. Building FSM manually is also feasible but it will be more error prone for several reasons. First, MSM can handle situation when the FSM is in a state that can't handle certain event but this event shouldn't be discarded. In this case the FSM can be in a `starting` or `stopping` state it might receive any event while in one of these states. Here it's handled by `Defer` function which does this for us. Second, the MSM does some compile time and run-time validation.

Reply:
Yea I think it makes sense to continue using it. It does seem pretty robust.

Is there a comment explaining the `none` and `defer` columns? Do those just mean that we don't make a transition?

Reply:
none in the target column means no transition
defer means that the event is queued and will be retriggered in the next state, so basically it's deferred until the next state transition